### PR TITLE
[APPS-2101] Add multibrand support with filtering

### PR DESCRIPTION
**Pull request by bjz**
:koala: 

cc. @zendesk/quokka, @adammw, @svizzari, @danielbreves 

It's not pretty, and it uses an internal API, but it works :)

Thanks goes to @adammw for finding that API spelunking.

multibrand2

### References

- [APPS-2101](https://zendesk.atlassian.net/browse/APPS-2101)

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -11,11 +11,13 @@
       // AJAX EVENTS
       'searchHelpCenter.done': 'searchHelpCenterDone',
       'searchWebPortal.done': 'searchWebPortalDone',
+      'getBrands.done': 'getBrandsDone',
       'getHcArticle.done': 'getHcArticleDone',
       'getSectionAccessPolicy.done': 'getSectionAccessPolicyDone',
       'settings.done': 'settingsDone',
 
       // DOM EVENTS
+      'change select[name="brand"]': 'processSearchFromInput',
```
Comment (princemaple) [inline app.js]: better use a class..

File: app.js [hunk 2]
```diff
@@ -123,6 +132,13 @@
       return data.agent_only || data.access_policy && data.access_policy.viewable_by !== 'everybody';
     },
 
+    getBrandsDone: function (data) {
+      var $select = this.$('select[name="brand"]');
```
Comment (princemaple) [inline app.js]: better use a class

File: app.js [hunk 3]
```diff
@@ -134,8 +150,17 @@
...
 
-    searchHelpCenterDone: function(data){
-      this.renderList(this.formatHcEntries(data.results));
+    searchHelpCenterDone: function(data) {
+      var brand = this.$('select[name="brand"] option:selected').val(); 
+      if (brand == 'any') {
+        this.renderList(this.formatHcEntries(data.results));
+      } else {
+        this.renderList(this.formatHcEntries(
+          _.filter(data.results, function (article) {
```
Comment (adammw) [inline app.js]: I was going to say that we can do the filtering via the API, but I kinda like this better, it'll be a lot more responsive (as long as we deal with pagination properly)

File: app.js [hunk 4]
```diff
@@ -123,6 +132,13 @@
       return data.agent_only || data.access_policy && data.access_policy.viewable_by !== 'everybody';
     },
 
+    getBrandsDone: function (data) {
+      var $select = this.$('select[name="brand"]');
+      _.each(data.brands, function(brand) {
+        $select.append('<option value="' + brand.id + '">' + brand.name + '</option>');
```
Comment (princemaple) [inline app.js]: I'd prefer using a template and use `rendarTemplate(brand.id, brand.name)`

File: app.js [hunk 5]
```diff
@@ -134,8 +150,17 @@
       if (this.isAgentOnlyContent(data)) { this.renderAgentOnlyAlert(); }
     },
 
-    searchHelpCenterDone: function(data){
-      this.renderList(this.formatHcEntries(data.results));
+    searchHelpCenterDone: function(data) {
+      var brand = this.$('select[name="brand"] option:selected').val();
```
Comment (princemaple) [inline app.js]: prefer class selector

File: templates/layout.hdbs [hunk 6]
```diff
@@ -9,6 +9,9 @@
   <div class="input-append custom-search">
     <input type="text" placeholder="{{t "layout.placeholder_text"}}">
     <button class="btn" type="button">{{t "layout.search_button"}}</button>
+    <select name="brand">
```
Comment (adammw) [inline templates/layout.hdbs]: Any chance we can hide this by default (ie a hidden class) and only show it when the brands API returns and there's more than one brand? The most common case will be accounts without multibrand where we shouldn't show anything.

**Comment by princemaple** [comment]
looks pretty good

File: app.js [hunk 7]
```diff
@@ -184,8 +209,8 @@
 
     formatHcEntries: function(result){
       var slicedResult = result.slice(0, this.numberOfDisplayableEntries());
-      var entries = _.inject(slicedResult, function(memo, entry){
-        var title = entry.name;
+      var entries = _.inject(slicedResult, function(memo, entry) {
+        var title = entry.brand_name + ": " + entry.name;
```
Comment (adammw) [inline app.js]: Similarly, would you be able to think of a way to do this conditionally, ie. not add this information if there's only one brand?

File: templates/layout.hdbs (hunk 6 above)
Comment (bjz) [inline templates/layout.hdbs]: Makes sense.

Comment (svizzari) [inline templates/layout.hdbs]: We could also use a Zendesk-styled select here.  https://github.com/zendesk/zendesk_app_framework/blob/master/lib%2Ftemplates%2Fzd_menu.hdbs.  You can use something like:
```js
{{zd_menu "select" { options: [{}, {}, ...] }}}
```
if you want to one-line the whole thing into a template then use `renderTemplate`

Comment (svizzari) [inline templates/layout.hdbs]: Related to above comment ^, you can update the menu after it has been created by doing something like:
```js
this.$('.brand-select').zdSelectMenu('setValue', [{}, {}, ...])
```

**Comment by bjz** [comment]
I need to address the case of having no brands - just a sec.

**Comment by bjz** [comment]
Ok, pretty sure I addressed all the critiques.

File: app.js [hunk 8]
```diff
@@ -123,6 +146,12 @@
       return data.agent_only || data.access_policy && data.access_policy.viewable_by !== 'everybody';
     },
 
+    getBrandsDone: function(data) {
```
Comment (adammw) [inline app.js]: Looks a little strange to me to see both a request.done callback and a promise-based callback. Do you think you could merge the two?

File: app.js [hunk 9]
```diff
@@ -91,9 +99,24 @@
         return this.initialize();
     },
 
+    isMultibrand: function() {
+      return this.brands.length > 1;
+    },
+
     initialize: function(){
-      if (_.isEmpty(this.ticket().subject()))
+      if (_.isEmpty(this.ticket().subject())) {
```
Comment (adammw) [inline app.js]: Nice.

File: app.js [hunk 10]
```diff
@@ -134,8 +163,17 @@
       if (this.isAgentOnlyContent(data)) { this.renderAgentOnlyAlert(); }
     },
 
-    searchHelpCenterDone: function(data){
-      this.renderList(this.formatHcEntries(data.results));
+    searchHelpCenterDone: function(data) {
+      var brand = this.$('.custom-search .brand-filter option:selected').val();
+      if (brand === undefined || brand === 'any') {
```
Comment (adammw) [inline app.js]: alternatively you could declare a variable results, set it initially to data.results, and then filter it if `brand && brand != 'any'` before rendering the variable. #codegolf

**Comment by svizzari** [comment]
@bjz I think you missed my comments about a Zendesk styled select completely because of a force push (or just because you addressed one of @adammw's comments).

**Comment by adammw** [comment]
I think the UI looks a little off with the combo box added there but it definitely satisfies a POC. :+1: from me.

**Comment by bjz** [comment]
Now I'm just trying to figure out the `zd_menu` handlebars helper.

**Comment by bjz** [comment]
:haircut:

Now using the lotus select widget. I updated the gif in the description showing this.

**Comment by adammw** [comment]
Looks good! :+1: still.

File: app.js (hunk 8 above)
Comment (bjz) [inline app.js]: This has been fixed! :)

File: app.js [hunk 11]
```diff
@@ -184,8 +219,9 @@
 
     formatHcEntries: function(result){
       var slicedResult = result.slice(0, this.numberOfDisplayableEntries());
-      var entries = _.inject(slicedResult, function(memo, entry){
-        var title = entry.name;
+      var entries = _.inject(slicedResult, function(memo, entry) {
+        var title = this.isMultibrand ? entry.brand_name + ": " + entry.name
```
Comment (princemaple) [inline app.js]: prefer
```js
var title = entry.name;
if (this.isMultibrand?) { title = entry.brand_name + title; }
```

**Comment by princemaple** [comment]
other than that ^ :+1:

File: app.js (hunk 11 above)
Comment (adammw) [inline app.js]: #playingcodegolf #don'tactuallyusethiscode
``js
var title = [entry.name];
if (this.isMultibrand) { 
  title.unshift(entry.brand_name);
}
memo.push({
  title: title.join(': ')
})
``
:p

Comment (princemaple) [inline app.js]: there must be one person between you and me that misunderstands codegolf... I thought it was aiming for as less code as possible?

Comment (adammw) [inline app.js]: Yeah, I just find string interpolation / concatenation especially ugly. Carry on... :walking:

**Pull request merged**